### PR TITLE
groupingby: add missing "discarded" counter

### LINKED ISSUE #1738: grouping-by-parser: missing discarded stats counter
start syslog-ng
send a line to input file:
<38>Feb 25 14:09:07 testhost testapp: first message

expected:
there is a discarded (0) line for this parser in syslog-ng-ctl stats result.

reality: it is missing

config:
@version: 3.12
@module "kvformat"
options {
    keep_hostname(yes);
    threaded(yes);
    stats_level(1);
};
source s_file_65caafa6dfb54a1bb92ceb937d7b90ed {
    file("/home/pzolee/var/testdb_working_dir0/d5312990-e083-405e-b549-55d3a31a63a2.txt" keep_timestamp(yes));
};
parser p_kv_parser_59ad1077e5ea4422835bce835f049063 {
    kv_parser(prefix('.SDATA.kvparser.'));
};
parser p_grouping_py_parser_43f6c545031e47eb9691c5b027ee802c {
    grouping_by(aggregate(value("MESSAGE", "${MESSAGE}@2"))
 key("$PID")
 timeout(1));
};
destination d_file_0b2313526d5646c78754234b11684b0a {
    file("/home/pzolee/var/testdb_working_dir0/14c15765-25b5-41a2-b03a-cd1adb591e5f.txt" flags(syslog-protocol));
};
log {
source(s_file_65caafa6dfb54a1bb92ceb937d7b90ed);
parser(p_kv_parser_59ad1077e5ea4422835bce835f049063);
parser(p_grouping_py_parser_43f6c545031e47eb9691c5b027ee802c);
destination(d_file_0b2313526d5646c78754234b11684b0a);

flags(flow-control);
};

### PR DESCRIPTION
**Pull request by bazsi**
This patch adds log_parser_init_method() to the init() call chain,
where it was missing. This fixes #1738, as well as any potential
missing initialization problem in the future.

Signed-off-by: Balazs Scheidler <balazs.scheidler@balabit.com>

### CODE REVIEW

**Comment by kira-syslogng** [comment]
Build SUCCESS, the tests were executed on test branch: master and test suite: functions

File: modules/dbparser/groupingby.c [hunk 1]
```diff
@@ -387,6 +387,9 @@ grouping_by_init(LogPipe *s)
   GroupingBy *self = (GroupingBy *) s;
   GlobalConfig *cfg = log_pipe_get_config(s);
 
+  if (!log_parser_init_method(s))
```
Comment (gaborznagy) [inline modules/dbparser/groupingby.c]: Should not grouping-by parser call StatefulParser's init method instead?

**Comment by bazsi** [comment]
it should. thanks for the comment.


-- 
Bazsi

**Comment by bazsi** [comment]
due to @gaborznagy's comment I have now added a broader refactor.

1) log_parser_deinit_method() is introduced, so there's something to call from stateful_parser_deinit()
2) this was added to all LogParser instances
3) stateful_parser_init/deinit is added
4) both db-parser and grouping-by are using the same init/deinit calls.

**Comment by kira-syslogng** [comment]
Build FAILURE, $TEST_DESC_COMMENT

**Comment by kira-syslogng** [comment]
Build SUCCESS, the tests were executed on test branch: master and test suite: functions

**Comment by pzoleex** [comment]
@kira-syslogng test this please test branch=pzolee-grouping-by-parser;

**Comment by kira-syslogng** [comment]
Build SUCCESS, the tests were executed on test branch: pzolee-grouping-by-parser and test suite: functions

**Pull request merged**